### PR TITLE
fix(commonjs): Support __esModule packages with a default export

### LINKED ISSUE #266: Undefined exported __moduleExports
- Rollup Plugin Name: @rollup/plugin-commonjs
- Rollup Plugin Version: 11.0.2
- Rollup Version: 1.32.1
- Operating System (or Browser): Arch Linux 
- Node Version: v12.16.1

### How Do We Reproduce?
Not really sure. Try to roll a mix of ES6 modules and CJS modules that require named export '__moduleExports' when your main package.js has "type": "module".

In my projectI was using snowpack (https://www.snowpack.dev/) with this configuration

/* snowpack.config.js */
module.exports = {
    namedExports: {
        sanctuary: ['__moduleExports'],
        xstream: ['__moduleExports'],
        'sanctuary-def': ['__moduleExports'],
        'sanctuary-type-identifiers': ['__moduleExports']
    },
    rollup: {
        plugins: [pluginNodeBuiltins(), pluginNodeGlobals()]
    }
}

### Expected Behavior
Working rolled web_modules

### Actual Behavior
Exported __moduleExports is undefined

# Proposed solution
Solved this way: https://github.com/FbN/plugins/commit/829219e6c0a4e9c3e458575d4cb9e14b12221477

### PR DESCRIPTION
**Pull request by FredKSchott**
## Rollup Plugin Name: `commonjs`

This PR contains:

- [X] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [X] yes
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [X] no - I consider this a bugfix, but I'll leave it up to the team how to treat this

List any relevant issue numbers: #266, https://www.pika.dev/npm/snowpack/discuss/367

### Description

This fixes #266, and is a follow up to this comment: https://github.com/rollup/plugins/issues/451#issuecomment-644459085

> unwrapExports is given the already processed Common.js module, where x.foo === module.exports.foo. We should only need to unwrap further if x was not compiled with __esModule support, in which case everything would still live on the default export. If __esModule exists, there's no need to unwrap.

> I can confirm this by removing Object.defineProperty(exports, "__esModule", {value: true}); from is-hotkey in my project, and watching it compile correctly.

My understanding is that this used to be needed for Common.js handling, but as of the latest changes in the latest version of Rollup/this plugin, this function call is no longer needed and is actually problematic, since unwrapping is happening elsewhere (and maybe the `unwrap` function can even be removed entirely, but would want feedback from a reviewer first)

### CODE REVIEW

File: packages/commonjs/test/test.js [hunk 1]
```diff
@@ -153,7 +153,7 @@ test('handles transpiled CommonJS modules', async (t) => {
   const fn = new Function('module', 'exports', code);
   fn(module, module.exports);
 
-  t.is(module.exports, 'foobar', code);
+  t.deepEqual(module.exports, { __esModule: true, default: 'foobar' }, code);
```
Comment (FredKSchott) [inline packages/commonjs/test/test.js:156]: considering this test as broken as written, and fixed by this PR

**Comment by lierdakil** [comment]
If I had to guess, #149 is the change that broke this. The caveat is, as far as I can tell, `syntheticNamedExports` has no provision for _both_ default and named exports in the same module (correct me if I'm wrong).

With `unwrapExports`, named exports get ignored. In case `unwrapExports` is removed, named exports will work, but the default export will not. This PR has the same issue.

**Comment by lukastaegert** [comment]
Looking at this, I wonder if removing `unwrapExports` altogether is the right move. But would need to check in what situations it actually applies. Before this change, it implemented the interop that extracted the default export from an ES module that was transpiled to CJS.
One question is what happens if a module is converted from ESM to CJS via Rollup and then back via this plugin, will this now preserve all exports?

**Comment by lierdakil** [comment]
As I said,

> With unwrapExports, named exports get ignored. In case unwrapExports is removed, named exports will work, but the default export will not.

That's because rollup will take `export default` at face value and try to use the exported object as the default export, instead of that object's `default` property.

So, no, probably not?

This could perhaps be fixed in rollup's core to tell it to look for default export in `default` property when `syntheticNamedExports` is `true`.

**Comment by lierdakil** [comment]
Here's a quick testcase: https://gist.github.com/lierdakil/8af74b2dc5a5aea76d443803f1b6f902
```js
export default 123;
export const x = 321;
```
is converted by rollup to CJS as
```js
Object.defineProperty(exports, '__esModule', { value: true });
var test = 123;
const x = 321;
exports.default = test;
exports.x = x;
```
If we then import this result from another module (both default and x):
```js
import val,{x} from './test-out.js'

console.log('val=', val)
console.log('x=', x)
```
this all gets bundled into:
```js
// ...
var testOut = createCommonjsModule(function (module, exports) {

Object.defineProperty(exports, '__esModule', { value: true });

var test = 123;
const x = 321;

exports.default = test;
exports.x = x;
});

var val = /*@__PURE__*/unwrapExports(testOut);

console.log('val=', val);
console.log('x=', val.x);
```
Now, whatever `unwrapExports` does or does not would be incorrect, the correct version would be
```js
console.log('val=', val.default);
console.log('x=', val.x);
```

For reference, commonjs-plugin v12 outputs the following:
```js
// ...
var val = unwrapExports(testOut);
var testOut_1 = testOut.x;

console.log('val=', val);
console.log('x=', testOut_1);
```
which will work correctly.

Side note, v13.0 also mishandles named exported functions: it calls those as a method on module, which binds `this` to the module instead of binding to current `this`, i.e. given
```js
const obj = { func: function() { console.log(this); } };
```
consider the difference between
```js
obj.func();
```
and
```js
const func = obj.func;
func();
```

**Comment by lukastaegert** [comment]
> This could perhaps be fixed in rollup's core to tell it to look for default export in default property when syntheticNamedExports is true

I think actually we need a solution along these lines, or maybe a slightly different mode that does this. Will need to think about this a little, hope to find some time in the next days to start writing up a larger sketch of the interop issues  and goals we have and how we might to tackle them because a lot of things are interacting here. Meanwhile, I would prefer to put this PR on hold until we have a slightly broader picture.

**Comment by lierdakil** [comment]
The issue is, commonjs-plugin v13 is kinda-sorta horribly broken, so maybe revert #427 or mark v13 as deprecated on npm while figuring this out?

**Comment by shellscape** [comment]
> The issue is, commonjs-plugin v13 is kinda-sorta horribly broken, so maybe revert #427 or mark v13 as deprecated on npm while figuring this out?

Maybe I'm having an off day, maybe it's a language barrier thing, but I _really_ don't appreciate the snark here. If you truly feel that the version is "horribly broken," you're welcome to use the prior version. Forking is always an option too. Regardless, please dial down the language - it's not helping.

**Comment by lierdakil** [comment]
To be clear, I did not mean to imply anything about any person with any relation to the project, only that the version in question can cause quite a bit of pain for the end-user. It introduces bugs that are not immediately apparent for reasonably common cases (i.e. "broken"). And it's not immediately obvious that it is doing that, which is, in my opinion, rather horrible.

I don't want to turn this into code of conduct discussion, but I will say that I don't really see the problem. English is not my native language, so some subtleties might elude me. If you want to discuss my use of language further, feel free to e-mail me (the address is in my GitHub profile)

> If you truly feel that the version is "horribly broken," you're welcome to use the prior version

I am currently, actually. But I've spent 6 hours debugging this on a new setup before I figured out what's going on and that I need to downgrade. You have to agree, it's far from an ideal user experience. v13 has several issues I've outlined above, so having it as the default for new setups is, let's call it, not optimal. Hence, the suggestion to revert the change or mark the version as deprecated.

**Comment by guybedford** [comment]
@FredKSchott so I wonder if we could just directly update this PR to entirely remove `unwrapExports`? Does that expose other compatibility issues?

**Comment by FredKSchott** [comment]
+1 for removing unwrapExports entirely if others are onboard.

Done in bfc81b4, all tests are passing with snapshots updated (see diff)

**Comment by guybedford** [comment]
(note this is still a breaking change though and should be a major - so perhaps land this along with any other interop PRs)

**Comment by danielgindi** [comment]
If you guys want to wait a few days for a major version, so I’ll get a little bit of time on my hands, I want to go over all of the interop code and see if there’s something we can do to satisfy all situations and be consistent.

**Comment by FredKSchott** [comment]
+1 for doing this the right way and not rushing it as long as there's bandwidth to do so in the near future. If it came down to it, I'd definitely recommend an incremental approach over never doing it at all :)

**Comment by marvinhagemeister** [comment]
Can confirm that this fixes a [similar issue](https://github.com/rollup/plugins/issues/400#issuecomment-651039247) I was running into this morning :+1:

**Comment by FredKSchott** [comment]
Bump! We've seen a couple more issues reported regarding this in Snowpack (ie https://www.pika.dev/npm/snowpack/discuss/421) and would love to get it fixed. Let me know if there's anything that I can do to help.

**Comment by lukastaegert** [comment]
Having created #481 I think this PR is likely a good stop-gap measure until we implemented a proper solution as outlined in #481. I think the only cases that are likely broken after this PR is when importing the default from a CJS module that was a transpiled ES module with named export mode, i.e. when the default export is on the `.default` property. However I expect this case to be rather rare in the wild especially since Rollup warns you when creating such a package. My recommendation is now to release this as a bug-fix patch release while we work through #481 and make that the next major.

**Comment by victoriafrench** [comment]
The removal of unwrapExports as a patch no less, is now breaking a bunch of builds. An example of one that breaks with it gone is `redux-subscriber`.

**Comment by TrySound** [comment]
PR title is very confusing. Looks like __esModule support is dropped here, not added. Draft-js plugins are broken for me after commonjs plugin upgrade.
https://github.com/rollup/plugins/pull/465

**Comment by shellscape** [comment]
@TrySound sounds like that should be tracked in another issue.

**Comment by simonhaenisch** [comment]
Yeah I agree with @TrySound that the PR title is confusing. This is more like "break __esModule packages with a default export so that named exports work" 😅 I kind of get the fix but interesting that a fix that's already known to break something else gets so much approval. Luckily there's already #501 which I hope has fixed this 👏

> > If you truly feel that the version is "horribly broken," you're welcome to use the prior version
> 
> I am currently, actually. But I've spent 6 hours debugging this on a new setup before I figured out what's going on and that I need to downgrade. You have to agree, it's far from an ideal user experience. v13 has several issues I've outlined above, so having it as the default for new setups is, let's call it, not optimal. Hence, the suggestion to revert the change or mark the version as deprecated.

Kind of agree with this. Just read the explanation in https://github.com/rollup/plugins/issues/497#issuecomment-657952505... v13.0.0 apparently broke a lot of packages because of `unwrapExports` so it would make sense to deprecate it. This fix (v13.0.1) removed `unwrapExports` which broke a few other packages like `firebase/messaging`, so it would also make sense to deprecate it. But then the latter issue persists for v13.0.2 and v14.0.0 as well, so you'd need to deprecate them as well, and that sounds a bit excessive 😅 

Deprecation doesn't mean that the version isn't installable anymore btw, just that you'll be able to show a deprecation message to devs (see [docs](https://docs.npmjs.com/deprecating-and-undeprecating-packages-or-package-versions#deprecating-a-single-version-of-a-package)), like

```
npm deprecate @rollup/plugin-commonjs@13.0.1 "This version breaks the default export of __esModule packages."
```

Could save some devs some time, or we just hope they'll find this thread 😄

BTW we've ran into this with `@stencil/core` which uses rollup to bundle the user's code, and therefore `@rollup/plugin-commonjs` is not a visible dependency to the end user anymore. Versions `1.16.x` and `1.17.0` of stencil are broken when using a dependency that's ESM with default export -> CJS  -> imported as ESM (pretty sure that's exactly what #501 fixed).

/cc @bitflower who had issues with stencil + `feathers-reduxify-authentication` because of this.

**Comment by lierdakil** [comment]
I've agonized about writing this for a while, but I think it's something that should be said. Apologies if it's grating to some people, and sorry for going off-topic, but I don't want to start a new thread just for this. Please don't take this as a personal attack on anyone, this is not my intention. I just think I'm noticing a pattern, I believe it should be pointed out, and I don't know how to sugar-coat it.

So I've decided that I'm done with rollup, at least for the foreseeable future. I won't be using it in any new projects, and will eventually phase it out in current ones. Why? Because when the dev team wilfully publishes code that is known to be broken, refuses to mark published code known to be broken as broken (i.e. deprecate the broken versions), and ignores or even shuts down criticism of that decision with "holier than thou" attitude, I believe it betrays a fundamental issue with internal culture. I'm all for supportive and positive social environment, but we have to remember that we, as developers, should strive to deliver a good product first and foremost, and I feel that rollup team lost sight of this simple truth, and are instead content to cheer each other on while introducing a change that fixes one thing while very obviously breaking another, instead of doing the responsible (in my opinion) thing -- rolling back the breaking changes, and going back to the drawing board. If I'm reading the situation correctly (which I might be not), we can expect more of the same down the line, and I won't be using software which I fully expect might cause issues with each release (even patch ones).

Granted, I'm a just some guy from the street, so I don't claim to know what's going on here. But this is what I've got from my brief interactions here and observing this repo for the last month or so. And of course I'm fully aware that the dev team doesn't owe me, or any of their users for that matter, anything, and they are free to do as they wish. What I wrote here, at the end of the day, is just an opinion piece, which you're free to agree or disagree with. But I hope at least that maintainers read it and reflect on it, and not dismiss it with a "good riddance, you horrible person"-type thought.

I really would prefer to avoid turning this into a flame war, so I will abstain from any further discussion in this thread. If you find I'm misunderstanding or misrepresenting something, and you fell you need to argue with me about it, my e-mail address is in my GitHub profile. You are free to respond here, of course, but I will try my best to not respond back.

I sincerely hope this comment will be food for thought, and not for indignation. Best wishes.

P.S. This sounds a lot harsher than I want it to, so a small addendum to hopefully soften this a little. Being an accidental maintainer myself, I know it can be a rather thankless job, so I want to say _thank you_ for your work. I might disagree with how things are handled here, and might avoid using rollup for pragmatic reasons, but it doesn't mean I don't appreciate the work you're putting in. So thank you, @lukastaegert, @shellscape. Terribly sorry if I'm forgetting someone.

**Comment by shellscape** [comment]
For the sake of full disclosure; I've deleted the comment by @lierdakil. It was, in every sense of the term, off-topic. The issues and PR threads on the repository are not a forum for airing grievances, proselytizing, rants, nor for declaring disdain or departure from the project. If during an unfortunate set of events one chooses to move on from Rollup, please do so without posting here. There are many other forums and mediums for doing so that are far more effective.

As this PR has gone off the rails, it's been locked from further comments. If you have an issue or bug that has not been addressed by other issues, please do open a new issue.

**Pull request merged**